### PR TITLE
feat: avoid updating relations in place

### PR DESCRIPTION
**Pull request by kian99**
## Description

This PR attempts to address https://github.com/juju/terraform-provider-juju/issues/235 which fails when the relation between Traefik and tls-certificates-operator charm is switched to be between Traefik and Vault - for the "certificates" interface. Because Traefik only allows a maximum of 1 relation for this interface and because our update logic first creates the new relation before destroying the old one, we run into this error when updating in place.

When writing a test to replicate this I started by using the juju-qa-dummy-(sink/source) charms. These don't enforce a maximum limit of 1 relation so we would expect this test to pass but the test still failed with the error,
```
        Error: Provider produced inconsistent result after apply

        When applying changes to juju_integration.this, provider
        "provider[\"registry.terraform.io/hashicorp/juju\"]" produced an unexpected
        new value: .id: was
        cty.StringVal("tf-test-integration-148028719398472653:one:source:two:sink"),
        but now
        cty.StringVal("tf-test-integration-148028719398472653:one:source:three:sink").

        This is a bug in the provider, which should be reported in the provider's own
```

It seems that although the new relation is created successfully, when Terraform runs a final `read()` operation to confirm the resources match what was just created, it is still seeing the old relation (because we were not waiting for its removal).

Normally we don't wait for resource removal in the `juju` client package, we do the wait in the `provider` package. So I removed the resource's `update()` method and marked all the fields as `requiresReplace()` because if we inspect the `Update()` logic it is doing a replacement **anyway**. Now, lo and behold the test passes.

Unfortunately it's not quite as simple as this. The update() logic did a bit more. If the relation was cross-model, the `update()` method would also remove the consumed "saas" application. Removing that is also not totally correct because you may have multiple applications related to a single saas. There are some more details about that available in https://github.com/juju/terraform-provider-juju/issues/473. 

After some discussion, the new plan to address is to make the changes to fix this in this PR, i.e.,
1. Handle the problem of multiple apps relating to a single "saas" by using app aliases (e.g. `juju consume <offer-url> <alias>`) such that each remote app that is created via an offer, creates a separate saas app. I.e. if the offer url is `admin/model.mysql` we alias the remote-app that will be created by adding a unique suffix and storing that suffix in Terraform's state. So the created remote-app becomes `mysql-<localApp>-<endpoint>`. And now the saas can safely be removed when the integration is destroyed because relations are 1-1 with remote-apps.
2. Ensure the above change is backwards compatible with existing deployments with documentation for edge cases.
3. Always update integrations by using `RequiresReplace` on all fields.
4. Remove saas apps when we remove integration that are cross-model.

Implementation notes:
- We don't need to store the app suffix separately, since we only set the suffix during creation. So existing deployments won't be affected.
- We can fetch the offer's consumed name from Juju, this ensures we are always using the correct name as recorded by Juju.

The PR now also addresses #473.

[JUJU-8528](https://warthogs.atlassian.net/browse/JUJU-8528)

<del>
What I propose we do for this bug and to simplify our modelling of relations:
1. Leave the bug unresolved in `main` i.e. v0.
6. Address the issue in `v1`, it doesn't necessarily have to be in the v1.0.0 release.
7. Handle the problem of multiple apps relating to a single "saas" by using app aliases (e.g. `juju consume <offer-url> <alias>`) such that each remote app that is created via an offer, creates a separate saas app. I.e. if the offer url is `admin/model.mysql` we alias the remote-app that will be created by adding a suffix and storing that suffix in Terraform's state. So the created remote-app becomes `mysql-qkxim`. And now the saas can safely be removed when the integration is destroyed because relations are 1-1 with remote-apps. Update: An alternative to the random suffix would be `<local-app>-<endpoint>`.
8. Ensure the above change is backwards compatible when upgrading an existing deployment (with documentation for any edge cases).
9. Always update integrations by using `RequiresReplace` on all fields.
10. Remove saas apps when we remove integration that are cross-model.
</del>

## QA Steps

I used the following plan to test a CMR and also validated that imports work as expected.
Plan 1:
```terraform
terraform {
  required_providers {
    juju = {
      source  = "registry.terraform.io/juju/juju" # (uses local provider repository)
      version = "0.22.0"
    }
  }
}

provider "juju" {
}

resource "juju_model" "this" {
  name = "source-model"

  cloud {
    name = "localhost"
    region = "localhost"
  }
}

resource "juju_application" "source" {
  name = "dummy-source"

  model = juju_model.this.name

  charm {
    name    = "juju-qa-dummy-source"
    channel = "latest/stable"
  }
  config = {
    "token" = "abc"
  }

  units = 1
}

resource "juju_offer" "source_offer" {
  model            = juju_model.this.name
  application_name = juju_application.source.name
  name             = "dummy-source"
  endpoints        = ["sink"]
}
```

Plan 2:
```terraform
terraform {
  required_providers {
    juju = {
      source  = "registry.terraform.io/juju/juju" # (uses local provider repository)
      version = "0.22.0"
    }
  }
}

provider "juju" {
}

resource "juju_model" "this" {
	name = "sink-model"

  cloud {
    name = "localhost"
    region = "localhost"
  }
}

resource "juju_application" "sink" {
  name = "dummy-sink"

  model = juju_model.this.name

  charm {
    name     = "juju-qa-dummy-sink"
    channel  = "latest/stable"
  }

  units = 1
}

data "juju_offer" "source-offer" {
  url = "admin/source-model.dummy-source"
}

resource "juju_integration" "sink_source_integration" {
  model = juju_model.this.name

  application {
    offer_url = data.juju_offer.source-offer.url
  }

  application {
    name     = juju_application.sink.name
    endpoint = "source"
  }

}
```

[JUJU-8528]: https://warthogs.atlassian.net/browse/JUJU-8528?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: docs-rtd/conf.py [hunk 1]
```diff
@@ -292,14 +292,15 @@
...
     # our own:
-    'sphinx_design',
+    "sphinx_design",
     # Make it possible to link to related RTD projects using their internal anchors
     # with, e.g., {external+ops:ref}`manage-configurations`:
-    'sphinxext.rediraffe',
+    "sphinxext.rediraffe",
     # Display an external link icon and open link in new tab:
     # new_tab_link_show_external_link_icon must also be set to True
-    'sphinx_new_tab_link',
```
Comment (luci1900) [inline docs-rtd/conf.py:301]: This kind of diff is still weirdly common in Python.

File: docs/resources/integration.md [hunk 2]
```diff
@@ -92,11 +92,44 @@ When creating this resource the `offer_url` property will show `(known after app
 This is due to an integration requiring a name/endpoint combination or an offer_url, but not both
 bits of data together.
 
+#### Cross-model relations
+
+From v0.23 of the provider, when integrating with an offer (i.e. when specifying the `offer_url`), the provider will create a remote application name with the format `<remote-app>-<local-app>-<interface>`. For example,
```
Comment (luci1900) [inline docs/resources/integration.md:97]: The parenthetical is a bit confusing to read with "provider" on both sides of it.

File: internal/juju/offers.go [hunk 3]
```diff
@@ -403,55 +471,45 @@ func (c offersClient) RemoveRemoteOffer(input *RemoveRemoteOfferInput) []error {
...
-		if url != input.OfferURL {
-			continue
-		}
-		offerName = v.OfferName
 	}
 
+	if offerName == "" {
+		return fmt.Errorf("remote-app %q not found in model", input.RemoteAppName)
+	}
+
```
[the remark this answers is not in the document]
Comment (kian99) [reply internal/juju/offers.go:494]: I think that would cause more issues for existing environments where it's not always the case that its 1:1 (remote-app:local-app).

[the remark this answers is not in the document]
Comment (kian99) [reply internal/juju/offers.go:494]: I'll merge so long, but happy to make any other follow-up adjustments today if needed.
